File: backend/app/schemas/merchant.py
"""Response schemas mirroring the iOS Merchant domain model.

Field names use snake_case internally; camelCase JSON aliases are emitted so the
iOS `Codable` layer decodes without any custom `CodingKeys`.
"""
from __future__ import annotations

import uuid
from datetime import datetime
from typing import List, Optional

from app.schemas.common import CamelModel


class BusinessHoursOut(CamelModel):
    summary: str
    closed_note: Optional[str] = None


class ReviewOut(CamelModel):
    id: uuid.UUID
    user_name: str
    rating: int
    content: str
    created_at: datetime
    payment_type: str
    payment_verified: bool
    purchased_product: Optional[str] = None


class PaymentVerificationOut(CamelModel):
    id: uuid.UUID
    payment_type: str
    succeeded_at: datetime
    note: Optional[str] = None


class MerchantOut(CamelModel):
    id: str
    name: str
    category: str

    latitude: float
    longitude: float

    address: str
    road_address: Optional[str] = None
    phone: Optional[str] = None

    # Populated by /nearby endpoint; otherwise omitted from JSON via exclusion.
    distance_meters: Optional[float] = None

    supports_onnuri: bool
    supports_local_currency: bool
    local_currency_name: Optional[str] = None
    supported_payment_types: List[str]

    products: List[str]
    business_hours: Optional[BusinessHoursOut] = None
    rating: float
    review_count: int

    market_name: Optional[str] = None
    description: Optional[str] = None
    last_verified_at: Optional[datetime] = None

    # Phase 13 Gate 3-B2 — 좌표 신뢰도 메타데이터.
    # iOS 는 이 필드로 개별 매장 마커 vs 시장 대표 마커 렌더링 분기 (docs/MAP_UX_TODO.md).
    location_source: Optional[str] = None
    location_precision: Optional[str] = None
    location_confidence: Optional[float] = None

    reviews: List[ReviewOut] = []
    recent_payments: List[PaymentVerificationOut] = []

File: LocalPayiOS/LocalPay/Models/Merchant.swift
import CoreLocation
import Foundation

/// 지도 위 가맹점 도메인 모델. CLAUDE.md §11.
///
/// 주의: MapKit / CoreLocation 의 클래스에 강하게 의존하지 않도록 좌표는 원시 Double 로 보관하고,
/// `coordinate` computed property 만 CoreLocation 을 참조한다.
struct Merchant: Identifiable, Hashable, Codable, Sendable {
    let id: String
    let name: String
    let category: MerchantCategory
    let latitude: Double
    let longitude: Double

    let address: String
    let roadAddress: String?
    let phone: String?

    /// 상세화면 헤더/카드에서 사용. 위치 권한 없을 때는 nil.
    var distanceMeters: Double?

    let supportsOnnuri: Bool
    let supportsLocalCurrency: Bool
    /// "안양사랑페이" 같은 지역화폐 브랜드명. 미지원이면 nil.
    let localCurrencyName: String?
    let supportedPaymentTypes: [PaymentType]

    let products: [String]
    let businessHours: BusinessHours?
    let rating: Double            // 0.0 ~ 5.0
    let reviewCount: Int

    let marketName: String?       // 예: "안양중앙시장"
    let description: String?
    let lastVerifiedAt: Date?

    let reviews: [Review]
    let recentPayments: [PaymentVerification]

    // Phase 13 Gate 3-C — 좌표 신뢰도 메타데이터 (docs/LOCATION_PRECISION.md).
    // 서버 응답 optional. Codable 자동 합성이 JSON 에 없으면 nil 로 매핑.
    // DummyMerchantSeed 등 로컬 코드는 nil 세 개 명시 전달.
    let locationSource: String?
    let locationPrecision: String?
    let locationConfidence: Double?

    var coordinate: CLLocationCoordinate2D {
        CLLocationCoordinate2D(latitude: latitude, longitude: longitude)
    }

    /// 시장 centroid 좌표 등 개별 매장이 아닌 근사 위치 (docs/MAP_UX_TODO.md).
    var isMarketLevelLocation: Bool {
        locationPrecision == "market_level"
    }

    /// 결제 수단 표시 헬퍼. Marker 색/아이콘 결정에 사용.
    var paymentBadge: PaymentBadgeKind {
        switch (supportsOnnuri, supportsLocalCurrency) {
        case (true, true):   return .both
        case (true, false):  return .onnuri
        case (false, true):  return .localCurrency
        case (false, false): return .none
        }
    }
}

/// 결제수단 조합 뱃지 종류.
enum PaymentBadgeKind: Hashable, Sendable {
    case onnuri
    case localCurrency
    case both
    case none
}
